src/core/dev/hw_queue_tx.cpp

```diff
@@ -1574,3 +1844,338 @@ int hw_queue_tx::modify_qp_ratelimit(struct xlio_rate_limit_t &rate_limit, uint3
+void hw_queue_tx::tx_task_completion_cb(doca_eth_txq_task_send *task_send, doca_data task_user_data,
+                                        doca_data ctx_user_data)
+{
+    mem_buf_desc_t *mem_buf = reinterpret_cast<mem_buf_desc_t *>(task_user_data.ptr);
+    hw_queue_tx *hw_tx = reinterpret_cast<hw_queue_tx *>(ctx_user_data.ptr);
+
+    hw_tx->return_doca_task(task_send);
+    hw_tx->handle_completion(mem_buf);
+}
+
+void hw_queue_tx::tx_task_lso_completion_cb(doca_eth_txq_task_lso_send *lso_task,
+                                            doca_data task_user_data, doca_data ctx_user_data)
+{
+    doca_lso_metadata *lso_metadata = reinterpret_cast<doca_lso_metadata *>(task_user_data.ptr);
+    hw_queue_tx *hw_tx = reinterpret_cast<hw_queue_tx *>(ctx_user_data.ptr);
+
+    hw_tx->return_doca_lso_task(lso_task);
+    hw_tx->handle_completion(lso_metadata->buff);
+    hw_tx->put_lso_metadata(lso_metadata);
+}
+
+void hw_queue_tx::tx_task_error_cb(doca_eth_txq_task_send *task_send, doca_data task_user_data,
+                                   doca_data ctx_user_data)
+{
+    mem_buf_desc_t *mem_buf = reinterpret_cast<mem_buf_desc_t *>(task_user_data.ptr);
+    hw_queue_tx *hw_tx = reinterpret_cast<hw_queue_tx *>(ctx_user_data.ptr);
+    doca_ctx_states ctx_state = DOCA_CTX_STATE_STOPPING;
+    doca_error_t rc_state = doca_ctx_get_state(hw_tx->m_doca_ctx_txq, &ctx_state);
+    ctx_state = ((ctx_state != DOCA_CTX_STATE_IDLE) ? ctx_state : DOCA_CTX_STATE_STOPPING);
+    if (rc_state != DOCA_SUCCESS || ctx_state != DOCA_CTX_STATE_STOPPING) {
+        PRINT_DOCA_ERR(__log_err,
+                       doca_task_get_status(doca_eth_txq_task_send_as_doca_task(task_send)),
+                       "TX Task Error");
+        // TODO DOCA: Add statistics for errors
+    }
+
+    hw_tx->return_doca_task(task_send);
+    hw_tx->handle_completion(mem_buf);
+}
+
+void hw_queue_tx::tx_task_lso_error_cb(doca_eth_txq_task_lso_send *lso_task,
+                                       doca_data task_user_data, doca_data ctx_user_data)
+{
+    doca_lso_metadata *lso_metadata = reinterpret_cast<doca_lso_metadata *>(task_user_data.ptr);
+    hw_queue_tx *hw_tx = reinterpret_cast<hw_queue_tx *>(ctx_user_data.ptr);
+    doca_ctx_states ctx_state = DOCA_CTX_STATE_STOPPING;
+    doca_error_t rc_state = doca_ctx_get_state(hw_tx->m_doca_ctx_txq, &ctx_state);
+    ctx_state = ((ctx_state != DOCA_CTX_STATE_IDLE) ? ctx_state : DOCA_CTX_STATE_STOPPING);
+    if (rc_state != DOCA_SUCCESS || ctx_state != DOCA_CTX_STATE_STOPPING) {
+        PRINT_DOCA_ERR(__log_err,
+                       doca_task_get_status(doca_eth_txq_task_lso_send_as_doca_task(lso_task)),
+                       "TX Task Error");
+        // TODO DOCA: Add statistics for errors
+    }
+
+    __log_func("tx_task_lso_error_cb, lso_task: %p, rc_state: %d, ctx_state: %d", lso_task,
+               rc_state, ctx_state);
+
+    hw_tx->return_doca_lso_task(lso_task);
+    hw_tx->handle_completion(lso_metadata->buff);
```

Comment:
Do we need to differentiate between successful and error completions?

Reply:
other than statistics - i dont think so... in prev flow we just called additional function called "log_cqe_error(cqe);" which prints a warning in case of err cqe.